Changed the train! function to support the syntax of the new explicit format.

Roeya opened the pull request:
Hi Lior
I have changed the syntax to the new explicit format of Flux 0.14.
I also changed the demo.jl script to use path in none OS dependency (joinpath) and relative to its location using @__DIR__

Roeya committed:
The train! function now uses the explicit format
data (varaible name) changed to train_data, make it more verbose
loss function no longer hides the model inside it and it does not get a Tupple you must explictly use m(x) as input for the y

Roeya committed:
Use paths relative to the demo.jl file loaction
Use join path to ensure OS compatability
loss function is no longer hide the model and uses tupple
Flux optimizer intitialized with new syntax
Call for train! refelcts the new explicit gradient claculation

Roeya committed:
Added missing dependecies of the examples

Roeya committed:
normalize path to remove backticks  - ".."

LiorSinai commented on examples/demo.jl:
```diff
@@ -15,7 +14,7 @@ using TransformersLite
 include("training.jl")
 
-path = "datasets\\amazon_reviews_multi\\en\\1.0.0\\"
+path = joinpath(@__DIR__, "..", "datasets", "amazon_reviews_multi", "en", "1.0.0")
```
The readme currently says change to this directory first, then the ".." won't be needed. So either this should be updated or the readme.

LiorSinai commented on examples/training.jl:
```diff
@@ -48,34 +48,34 @@ function split_validation(X::AbstractVector, Y::AbstractVector, frac=0.1; rng=Ra
 end
 
-function train!(loss, ps, data, opt, val_data; n_epochs=10)
+function train!(loss, model, train_data, opt, val_data; n_epochs=10)
```
Think opt should be called opt_state instead

Roeya replied:
done

LiorSinai commented on examples/training.jl:
```diff
@@ -48,34 +48,34 @@ function split_validation(X::AbstractVector, Y::AbstractVector, frac=0.1; rng=Ra
-            gs = gradient(ps) do
-                loss(Xy)
+        progress = Progress(length(train_data); desc="epoch $epoch/$n_epochs")
+        epoch_loss = 0.0
```
This epoch_loss is just left hanging here. But it might be worthwhile calculating it as the mean or sum of the batch losses.

To get the batch loss `val` do:
```
val, grads = Flux.withgradient(model) do m
    loss(m(Xy[1]), Xy[2])
end
```

Roeya replied:
done, added more loss info during progress reporting

LiorSinai reviewed (changes requested):
Overall looks good, just a few small comments.

Could you also add a compat in the project.toml for Flux > 0.13.9 since the main goal is to use new Flux syntax. Think that is also enough of a breaking change to justify changing the version to 0.2.0

LiorSinai commented on examples/demo.jl:
```diff
@@ -120,31 +119,31 @@ hyperparameters["model"] = "$(typeof(model).name.wrapper)"
 
 ## Training 
-opt = ADAM()
+opt = Flux.setup(Flux.Adam(), model)
```
`opt` should be called `opt_state`

Roeya replied:
done

Roeya committed:
Add mssing depedency (when running form command line)

Roeya committed:
Move back the TransformerClassifier after correcting trainable with named tupples

Roeya committed:
Converted to named tupple to suppprt the new trainable way

Roeya committed:
change opt to opt_state, remark code nt required (for Chain)

Roeya committed:
Added the batch loss and mean loss to progress

Roeya committed:
Missing dependecy

Roeya committed:
downgrade Flux to 0.13.9

Roeya committed:
fix bug in model show that required specifi handle of FlattenLayer

Roeya committed:
To support Flux 0.13.9 we need a workaround for a bug in Flux.setup
Se wee added Optimisers v0.2.15 and use Optimisers.setup
The problem was corrected at Flux 0.13.11
-Method error when using Flux.setup with Embedding layer (#2144)

Roeya committed:
set depndecies

LiorSinai commented on examples/demo.jl:
```diff
@@ -120,31 +121,29 @@ hyperparameters["model"] = "$(typeof(model).name.wrapper)"
-## Training 
-opt = ADAM()
+## Training (we use Optimisers instead of Flux due to bug in Flux below v0.13.11)
+opt_state = Optimisers.setup(Optimisers.Adam(), model)
```
The workaround is good but small question: why not target 0.13.11 instead?

My original idea was for 0.13.9 because that would allow the new syntax while having the broadest compatibility. Version 0.13.11 is only 2 patch versions ahead, so it won't make a big difference.

Version 0.14 is a different story. It has several breaking changes and requires the new Julia 1.9. A few of these are with package management - I see you encountered those errors in another commit. So I am hesitant to upgrade to it just yet.

Roeya replied:
Just followed your request to be compatible with 0.13.9 as I thought there is some reason for this specific version.
